Source/WebCore/xml/DOMParser.cpp

```diff
@@ -46,11 +51,30 @@ ExceptionOr<Ref<Document>> DOMParser::parseFromString(const String& string, cons
     auto document = DOMImplementation::createDocument(contentType, nullptr, m_settings, URL { });
     if (m_contextDocument)
         document->setContextDocument(*m_contextDocument.get());
+    OptionSet<ParserContentPolicy> parsingOptions;
     if (options.includeShadowRoots && document->settings().declarativeShadowDOMInDOMParserEnabled())
-        document->setParserContentPolicy({ ParserContentPolicy::AllowScriptingContent, ParserContentPolicy::AllowPluginContent, ParserContentPolicy::AllowDeclarativeShadowDOM });
+        parsingOptions = { ParserContentPolicy::AllowScriptingContent, ParserContentPolicy::AllowPluginContent, ParserContentPolicy::AllowDeclarativeShadowDOM };
     else
-        document->setParserContentPolicy({ ParserContentPolicy::AllowScriptingContent, ParserContentPolicy::AllowPluginContent });
-    document->setContent(string);
+        parsingOptions = { ParserContentPolicy::AllowScriptingContent, ParserContentPolicy::AllowPluginContent };
+    document->setParserContentPolicy(parsingOptions);
+    bool usedFastPath = false;
+    if (contentType == "text/html"_s) {
+        auto fragment = DocumentFragment::create(document);
+        auto body = HTMLBodyElement::create(document);
+        usedFastPath = tryFastParsingHTMLFragment(string, document, fragment, body, parsingOptions);
+        if (LIKELY(usedFastPath)) {
+            auto html = HTMLHtmlElement::create(document);
+            document->appendChild(html);
+            auto head = HTMLHeadElement::create(document);
+            html->appendChild(head);
+            auto result = body->appendChild(fragment);
+            if (UNLIKELY(result.hasException()))
+                return result.releaseException();
+            html->appendChild(body);
```

Comment:
Hm... This will traverse all of the subtree under body again. Have we tried always creating html & head ahead of calling tryFastParsingHTMLFragment?

Reply:
I haven't yet because it requires changing the fast parser so it doesn't generate a DocumentFragment anymore. I will investigate doing this in a follow-up though as odds are it will speed things up further.